File: frontend/src/types/equipment.ts
import { z } from "zod";

export const sensorSchema = z.object({
  id: z.string().optional(),
  sensor_type: z.string().min(1, "Sensor type is required"),
  mounting_location: z.string().min(1, "Mounting location is required"),
  orientation: z.string().min(1, "Orientation is required"),
  mounting_method: z.string().optional().nullable(),
  sensitivity: z.coerce.number().optional().nullable(),
  sensitivity_unit: z.string().optional().nullable(),
  sampling_rate: z.string().optional().nullable(),
  sampling_rate_custom: z.coerce.number().int().optional().nullable(),
  frequency_range: z.string().optional().nullable(),
  frequency_range_custom_min: z.coerce.number().int().optional().nullable(),
  frequency_range_custom_max: z.coerce.number().int().optional().nullable(),
  is_active: z.boolean().default(true),
});

/**
 * A number the user may simply not know yet.
 *
 * `z.coerce.number()` turns an empty input into `0`, so a blank optional field
 * that also carries `.positive()` fails validation — and react-hook-form
 * swallows a rejected submit, which is what made Save look dead when the
 * rotating-component fields were left empty. Blank, null, and unparseable
 * input all collapse to `null` here, before any numeric check runs.
 */
function optionalNumber(refine?: (schema: z.ZodNumber) => z.ZodNumber) {
  const base = z.coerce.number();
  return z.preprocess((value) => {
    if (value === "" || value === null || value === undefined) return null;
    if (typeof value === "number" && Number.isNaN(value)) return null;
    return value;
  }, (refine ? refine(base) : base).nullable().optional());
}

export const equipmentSchema = z.object({
  // Location & Hierarchy
  plant_name: z.string().optional().nullable(),
  area: z.string().optional().nullable(),
  line: z.string().optional().nullable(),

  // Asset Identification
  machine_name: z.string().optional().nullable(),
  machine_id: z.string().optional().nullable(),
  machine_type: z.string().optional().nullable(),
  machine_criticality: z.string().optional().nullable(),
  manufacturer: z.string().optional().nullable(),
  model: z.string().optional().nullable(),
  serial_number: z.string().optional().nullable(),

  // Mechanical Details
  rated_power_kw: z.coerce.number().positive().optional().nullable(),
  rated_rpm: z.coerce.number().int().positive().optional().nullable(),
  drive_type: z.string().optional().nullable(),
  load_type: z.string().optional().nullable(),
  foundation_type: z.string().optional().nullable(),
  coupling_details: z.string().optional().nullable(),

  // Rotating Components — every field here is optional: equipment saves with
  // the whole Bearing Details / Rotating Components step left blank.
  bearing_details: z.string().optional().nullable(),
  bearing_number_de: z.string().optional().nullable(),
  bearing_number_nde: z.string().optional().nullable(),
  // Bearing ID in the fault-frequency catalogue, set when the fitted bearing is
  // matched against it. Null means the bearing is recorded as free text only.
  bearing_de_catalog_id: optionalNumber((n) => n.int().positive()),
  bearing_nde_catalog_id: optionalNumber((n) => n.int().positive()),
  gearbox_ratio: optionalNumber((n) => n.positive()),
  gear_teeth: optionalNumber((n) => n.int().positive()),
  motor_pole_count: optionalNumber((n) => n.int()),
  fan_blades: optionalNumber((n) => n.int().positive()),
  pump_vanes: optionalNumber((n) => n.int().positive()),
  direction_of_rotation: z.string().optional().nullable(),

  // Operating Conditions
  operating_speed_min: z.coerce.number().int().optional().nullable(),
  operating_speed_max: z.coerce.number().int().optional().nullable(),
  load_range_min: z.coerce.number().optional().nullable(),
  load_range_max: z.coerce.number().optional().nullable(),
  normal_operating_load: z.coerce.number().optional().nullable(),
  process_details: z.string().optional().nullable(),
  operating_environment: z.array(z.string()).optional().nullable(),

  // Lubrication & Maintenance
  lubrication_type: z.string().optional().nullable(),
  installation_date: z.string().optional().nullable(),
  last_maintenance_date: z.string().optional().nullable(),
  maintenance_notes: z.string().optional().nullable(),

  // AI Readiness
  asset_status: z.string().optional().nullable(),
  machine_train_configured: z.boolean().optional(),
  bearing_database_mapped: z.boolean().optional(),
  operating_mode_configured: z.boolean().optional(),

  // Sensors
  sensors: z.array(sensorSchema).optional().default([]),
});

export type EquipmentFormData = z.infer<typeof equipmentSchema>;
export type SensorFormData = z.infer<typeof sensorSchema>;

export interface EquipmentListItem {
  id: string;
  plant_name: string;
  area: string;
  line: string;
  machine_name: string;
  machine_id: string;
  machine_type: string;
  machine_criticality: string;
  manufacturer?: string;
  asset_status?: string;
  equipment_image_path?: string;
  created_at: string;
}

export interface EquipmentOut extends EquipmentFormData {
  id: string;
  equipment_image_path?: string;
  created_at: string;
  updated_at: string;
}

export interface AIReadiness {
  equipment_id: string;
  score_percent: number;
  machine_train_configured: boolean;
  asset_status_set: boolean;
  sensor_coverage: boolean;
  bearing_database_mapped: boolean;
  operating_mode_configured: boolean;
}

export interface PaginatedEquipment {
  total: number;
  page: number;
  page_size: number;
  items: EquipmentListItem[];
}

export const CRITICALITY_COLORS: Record<string, string> = {
  Low: "bg-machine-healthy/10 text-machine-healthy border border-machine-healthy/25",
  Medium: "bg-machine-warning/10 text-machine-warning border border-machine-warning/25",
  High: "bg-machine-critical/10 text-machine-critical border border-machine-critical/25",
  Critical: "bg-machine-critical/15 text-machine-critical border border-machine-critical/30",
};

export const CRITICALITY_DOT: Record<string, string> = {
  Low: "bg-machine-healthy",
  Medium: "bg-machine-warning",
  High: "bg-machine-critical",
  Critical: "bg-machine-critical",
};

export const ASSET_STATUS_COLORS: Record<string, string> = {
  Active: "bg-machine-healthy/10 text-machine-healthy border border-machine-healthy/25",
  "Under Maintenance": "bg-machine-warning/10 text-machine-warning border border-machine-warning/25",
  Inactive: "bg-machine-offline/10 text-machine-offline border border-machine-offline/25",
  Offline: "bg-machine-offline/10 text-machine-offline border border-machine-offline/25",
};

File: backend/app/schemas/equipment.py
from __future__ import annotations
from typing import List, Optional
from datetime import date, datetime
from decimal import Decimal
from uuid import UUID
from pydantic import BaseModel, Field, field_validator


class SensorConfigBase(BaseModel):
    sensor_type: str
    mounting_location: str
    orientation: str
    mounting_method: Optional[str] = None
    sensitivity: Optional[Decimal] = None
    sensitivity_unit: Optional[str] = None
    sampling_rate: Optional[str] = None
    sampling_rate_custom: Optional[int] = None
    frequency_range: Optional[str] = None
    frequency_range_custom_min: Optional[int] = None
    frequency_range_custom_max: Optional[int] = None
    is_active: bool = True
    device_id: Optional[str] = Field(
        default=None,
        max_length=64,
        description="Edge device identifier (MAC-style), e.g. 11:AA:BB:CC:DD:EE",
    )


class SensorConfigCreate(SensorConfigBase):
    pass


class SensorConfigUpdate(BaseModel):
    sensor_type: Optional[str] = None
    mounting_location: Optional[str] = None
    orientation: Optional[str] = None
    mounting_method: Optional[str] = None
    sensitivity: Optional[Decimal] = None
    sensitivity_unit: Optional[str] = None
    sampling_rate: Optional[str] = None
    sampling_rate_custom: Optional[int] = None
    frequency_range: Optional[str] = None
    frequency_range_custom_min: Optional[int] = None
    frequency_range_custom_max: Optional[int] = None
    is_active: Optional[bool] = None
    device_id: Optional[str] = Field(default=None, max_length=64)


class SensorConfigOut(SensorConfigBase):
    id: UUID
    equipment_id: UUID
    created_at: datetime

    class Config:
        from_attributes = True


class EquipmentBase(BaseModel):
    # Location & Hierarchy
    plant_name: str = ""
    area: str = ""
    line: str = ""

    # Asset Identification
    machine_name: str = ""
    machine_id: Optional[str] = None
    machine_type: str = ""
    machine_criticality: str = ""
    manufacturer: Optional[str] = None
    model: Optional[str] = None
    serial_number: Optional[str] = None

    # Mechanical Details
    rated_power_kw: Optional[Decimal] = None
    rated_rpm: Optional[int] = None
    drive_type: Optional[str] = None
    load_type: Optional[str] = None
    foundation_type: Optional[str] = None
    coupling_details: Optional[str] = None

    # Rotating Components
    bearing_details: Optional[str] = None
    bearing_number_de: Optional[str] = None
    bearing_number_nde: Optional[str] = None
    gearbox_ratio: Optional[Decimal] = None
    gear_teeth: Optional[int] = None
    motor_pole_count: Optional[int] = None
    fan_blades: Optional[int] = None
    pump_vanes: Optional[int] = None
    direction_of_rotation: Optional[str] = None

    # Operating Conditions
    operating_speed_min: Optional[int] = None
    operating_speed_max: Optional[int] = None
    load_range_min: Optional[Decimal] = None
    load_range_max: Optional[Decimal] = None
    normal_operating_load: Optional[Decimal] = None
    process_details: Optional[str] = None
    operating_environment: Optional[List[str]] = None

    # Lubrication & Maintenance
    lubrication_type: Optional[str] = None
    installation_date: Optional[date] = None
    last_maintenance_date: Optional[date] = None
    maintenance_notes: Optional[str] = None

    # AI Readiness
    asset_status: Optional[str] = "Active"
    machine_train_configured: Optional[bool] = False
    bearing_database_mapped: Optional[bool] = False
    operating_mode_configured: Optional[bool] = False

    @field_validator("machine_id", mode="before")
    @classmethod
    def normalize_machine_id(cls, v):
        return None if v == "" else v


class EquipmentCreate(EquipmentBase):
    sensors: Optional[List[SensorConfigCreate]] = []


class EquipmentUpdate(BaseModel):
    plant_name: Optional[str] = None
    area: Optional[str] = None
    line: Optional[str] = None
    machine_name: Optional[str] = None
    machine_id: Optional[str] = None
    machine_type: Optional[str] = None
    machine_criticality: Optional[str] = None
    manufacturer: Optional[str] = None
    model: Optional[str] = None
    serial_number: Optional[str] = None
    rated_power_kw: Optional[Decimal] = None
    rated_rpm: Optional[int] = None
    drive_type: Optional[str] = None
    load_type: Optional[str] = None
    foundation_type: Optional[str] = None
    coupling_details: Optional[str] = None
    bearing_details: Optional[str] = None
    bearing_number_de: Optional[str] = None
    bearing_number_nde: Optional[str] = None
    gearbox_ratio: Optional[Decimal] = None
    gear_teeth: Optional[int] = None
    motor_pole_count: Optional[int] = None
    fan_blades: Optional[int] = None
    pump_vanes: Optional[int] = None
    direction_of_rotation: Optional[str] = None
    operating_speed_min: Optional[int] = None
    operating_speed_max: Optional[int] = None
    load_range_min: Optional[Decimal] = None
    load_range_max: Optional[Decimal] = None
    normal_operating_load: Optional[Decimal] = None
    process_details: Optional[str] = None
    operating_environment: Optional[List[str]] = None
    lubrication_type: Optional[str] = None
    installation_date: Optional[date] = None
    last_maintenance_date: Optional[date] = None
    maintenance_notes: Optional[str] = None
    asset_status: Optional[str] = None
    machine_train_configured: Optional[bool] = None
    bearing_database_mapped: Optional[bool] = None
    operating_mode_configured: Optional[bool] = None


class EquipmentOut(EquipmentBase):
    id: UUID
    equipment_image_path: Optional[str] = None
    sensors: List[SensorConfigOut] = []
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


class EquipmentListItem(BaseModel):
    id: UUID
    plant_name: str
    area: str
    line: str
    machine_name: str
    machine_id: Optional[str] = None
    machine_type: str
    machine_criticality: str
    manufacturer: Optional[str] = None
    asset_status: Optional[str] = None
    equipment_image_path: Optional[str] = None
    created_at: datetime

    class Config:
        from_attributes = True


class AIReadinessOut(BaseModel):
    equipment_id: UUID
    score_percent: int
    machine_train_configured: bool
    asset_status_set: bool
    sensor_coverage: bool
    bearing_database_mapped: bool
    operating_mode_configured: bool


class PaginatedEquipment(BaseModel):
    total: int
    page: int
    page_size: int
    items: List[EquipmentListItem]
